_posts/2021-01-05-git-rebase-merge.md

```diff
@@ -0,0 +1,171 @@
+
+Let's understand the working of Git `rebase` by looking at history with a topic branch off another topic branch. 
+
+Let's say we have branched a `feature1` branch from the mainline , and added some feature1-side functionality to our project, and then made a commit. Now, we branch off the `feature2` branch to make some feature2-side changes. Finally, we go back to the `feature1` branch and commit a few more changes:
+
+![Git Rebase Example](/assets/img/posts/git-rebase-merge/git-rebase-working1.png)
+
+Now suppose that we have decided to merge the feature2-side changes to the mainline for the release, but we also want to hold the feature1-side changes until they are tested further. 
+
+With `git rebase`, we can "replay" the changes in the `feature2` branch (that are not in the `feature1` branch, i.e. C8 and C9), and then replay them on the main branch by using the –onto option of git rebase, We have to specify all the three branches names in this case because we are holding the changes from feature1 branch while replaying them in the main branch from feature2 branch:
+
+```
+git rebase --onto main feature1 feature2
+```
+
+It gives us a bit complex but a pretty cool result:
+
+![Git Rebase Example](/assets/img/posts/git-rebase-merge/git-rebase-working2.png)
+
+
+Now it's time to fast forward our main branch. Fast forward is simply a unique instance of git rebase in which we are moving the tip of a branch to the latest commit. In our case, we want to move the tip of the main branch forward so it points to the latest commit of our feature2 branch.
+
+We will use the following commands to do this:
+
+```
+git checkout main
+
+git merge feature2
+```
+In simple words, fast forwarding `main` to the `feature2` branch means that previously the HEAD pointer for main branch was at 'C6' but after the the above command it fast forwards the main branch's HEAD pointer to the feature2 branch.
+
+![Git Rebase Done](/assets/img/posts/git-rebase-merge/git-rebase-working3.png)
+
+Let's understand the concept of Interactive Rebasing, Interactive rebasing helps us to clean up the commits before they are moved on to another branch. It is done to make the project history simpler and easier to read, also helping to avoid any future conflicts among the commits. 
+Suppose we have several commits, and somehow we want to modify them during the rebase, we can do this by passing an ‘-i’, or ‘interactive’ to the original ‘git rebase’ command.
+```
+git rebase -i origin/main
+```
+
+In this way, we have successfully invoked interactive rebase on all the commits we have pushed so far.
+
+
+## Git Rebase vs Git Merge
+
+Now let's go through the difference between `git rebase` and `git merge`. 
+
+When we merge, the branch we merge into will always preserve the ancestry of each commit history by generating a new commit:
+
+![Git Merge Preserving commit history](/assets/img/posts/git-rebase-merge/git-merge-history.png)
```

Comment:
Sorry, one more thing that I overlooked previously: I don't understand this diagram. Maybe I would understand it if the text explained it a bit more. The text only says, that "a merge preserves the commit ancestry", but doesn't explain how that relates to the image. Can you please explain some more in the paragraph below this image?

I guess the red commit is the merge commit? Just point it out in the text. What is the golden commit? Why don't we have a golden commit in the next image (the re-writing one)?

Reply:
Sure I will mention it in the paragraph, basically yes the red commit is the merge commit and to be able to merge we need a common base and the golden commit is the common base.
it is not present for the Git Rebase diagram because there is no need for a common as the commit history gets rewritten and not preserved for Git Rebase

Reply:
I have pushed a commit for these changes